Added new feature (new classes) to parse and load pdfs using PyMuPDF from bytes buffer

esiiol opened the pull request:
Hi, here is the required info for this pull request.

Description: 
I've added a new feature (new classes) to parse and load pdfs using PyMuPDF from bytes buffer. This is important when the PDF is not a file but is a bytes buffer in memory. Especially, it's useful when working and using Langchain with Spark and Databricks Autoloader. 
I've also added a unit test.

Dependencies: 
No new dependencies is required. Just io library which is in the native python packaging.

esiiol committed:
Added BaseBytesParser

this class is the base parser for bytes buffers.

esiiol committed:
Added PyMuPDFBytesParser

This class is the parser for `PDF` bytes using `PyMuPDF`

esiiol committed:
Added PyMuPDFBytesLoader

This class loads `PDF` byte buffers using `PyMuPDF`

esiiol committed:
Added new unit_tests for the new feature i.e. PDF bytes buffer parser

eyurtsev commented on libs/langchain/langchain/document_loaders/base.py:
```diff
@@ -93,3 +94,44 @@ def parse(self, blob: Blob) -> List[Document]:
         return list(self.lazy_parse(blob))
+
+
+class BaseBytesParser(ABC):
```
Hi @esiiol thank you for the contribution -- take a look if this can be used for your use case: 

Blob object which supports loading entirely from memory using bytes objects: https://github.com/langchain-ai/langchain/blob/68599d98c20b1e3fbdabaef1b1fbe54cd06b98a4/libs/langchain/langchain/document_loaders/blob_loaders/schema.py#L21-L21

https://github.com/langchain-ai/langchain/blob/68599d98c20b1e3fbdabaef1b1fbe54cd06b98a4/libs/langchain/langchain/document_loaders/base.py#L57-L57


And implementations of pdf parsers

https://github.com/langchain-ai/langchain/blob/68599d98c20b1e3fbdabaef1b1fbe54cd06b98a4/libs/langchain/langchain/document_loaders/parsers/pdf.py#L253-L253

esiiol replied:
Hi @eyurtsev thanks for your response. That's absolutely right. I'd missed them. I'll update the code accordingly and commit to the PR.

esiiol committed:
linting done

esiiol committed:
Update: added new feature for PyMuPDF to read from bytes buffer

deleted the new class that was  added before
modified the BasePDFLoader and PyMuPDFLoader

esiiol committed:
update: add new feature for PyMuPDFParser to read bytes

deleted the new class that was added and updated PyMuPDFParser

esiiol committed:
Delete unnecessary class I added before called BaseBytesParser

esiiol committed:
Update: updated unit tests to test new feature

esiiol committed:
tested, formatted and linted

esiiol committed:
Merge branch 'master' into master

eyurtsev commented on libs/langchain/langchain/document_loaders/pdf.py:
```diff
@@ -345,7 +353,10 @@ def load(self, **kwargs: Any) -> List[Document]:
+        if isinstance(self.file_path, str):
+            blob = Blob.from_path(self.file_path)
+        elif isinstance(self.file_path, BytesIO):
+            blob = Blob.from_data(self.file_path.read())
```
blob is not defined on the else branch -- could you raise a NotImplementedError -- useful since code is used interactively from notebooks so users will not be running static type checking in that environment

tevslin replied:
none of this code in pdf,py is mine. It's not one of the modules I chnaged

esiiol replied:
Done!

eyurtsev commented on libs/langchain/tests/unit_tests/document_loaders/parsers/test_pdf_parsers.py:
```diff
@@ -53,6 +54,51 @@ def _assert_with_parser(parser: BaseBlobParser, splits_by_page: bool = True) ->
         assert int(metadata["page"]) == 0
 
 
+def _assert_with_bytes_parser(
```
Could we pull out the dependent code into standalone function and have it all run from the same standard tester, so by default all parsers have to be verified to work with it?

tevslin replied:
I may be misunderstanding what you are asking. I did write a standalone unit test for it. Passing the standard test should work - eg, it should be able to read the test doc correctly - but that won't test its ability to extract email-like metadata since there isn't any in the test doc. But, if for completeness, this test should be there as well, glad to add it,

tevslin replied:
> Could we pull out the dependent code into standalone function and have it all run from the same standard tester, so by default all parsers have to be verified to work with it?

should this class be packaged in pypdf with the other parsers?

tevslin commented on libs/langchain/langchain/document_loaders/pdf.py:
```diff
@@ -66,16 +66,23 @@ class BasePDFLoader(BaseLoader, ABC):
         self.web_path = None
         self.headers = headers
+
+        if isinstance(self.file_path, BytesIO):
```
I didn't modify pdf,py

esiiol replied:
I've removed it

tevslin commented on libs/langchain/langchain/document_loaders/parsers/pdf.py:
```diff
@@ -206,28 +206,31 @@ def lazy_parse(self, blob: Blob) -> Iterator[Document]:
-                )
-                for page in doc
-            ]
+            if isinstance(blob.data, bytes):
```
I didn't modify pdf.py

esiiol replied:
because when the data is coming from the memory (i.e. Blob.from_data(self.file_path.read())) fitz.open(stream) should be used and when it's coming from a file (i.e. blob = Blob.from_data(self.file_path.read())) fitz.open(filename) should be used.

esiiol committed:
Update based on the comments in the PR

removed return from BasedPDFLoader.__init__
added ImplementationError in PyMuPDFLoader.load

esiiol commented:
I've updated the code according to the comments. However, linting was giving me some errors. Also, didn't re-implement the tester as I've seen another PR discussion on it.

hwchase17 commented:
closing due to inactivity and merge conflicts
